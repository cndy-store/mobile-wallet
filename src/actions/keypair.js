import { getItem, setItem, removeItem } from '../lib/keypairStorage';
import { StellarSdk } from '../lib/stellar';
const { Keypair, StrKey } = StellarSdk;

const STORAGE_KEY = 'STELLAR_SECRET';

export const KEYPAIR_STARTS_PROCESSING = 'KEYPAIR_STARTS_PROCESSING';
export const KEYPAIR_ERROR = 'KEYPAIR_ERROR';
export const KEYPAIR_LOAD = 'KEYPAIR_LOAD';
export const KEYPAIR_SAVE = 'KEYPAIR_SAVE';
export const KEYPAIR_REMOVE = 'KEYPAIR_REMOVE';

const chomp = str => str.replace(/\n|\r| /g, '');
const isValidSecret = secret => StrKey.isValidEd25519SecretSeed(secret);

export const keypairProcessing = bool => ({
  type: KEYPAIR_STARTS_PROCESSING,
  isProcessing: bool
});

export const keypairErrored = ({ error }) => ({
  type: KEYPAIR_ERROR,
  error
});

export const keypairLoadSucceded = ({ secret }) => {
  const keypair = secret ? Keypair.fromSecret(secret) : null;

  return {
    type: KEYPAIR_LOAD,
    keypair: keypair,
    error: null
  };
};

export const keypairSaveSucceded = ({ secret }) => {
  const keypair = Keypair.fromSecret(secret);

  return {
    type: KEYPAIR_SAVE,
    keypair: keypair,
    error: null
  };
};

export const keypairRemoveSucceded = () => ({
  type: KEYPAIR_REMOVE,
  keypair: null,
  error: null
});

export const loadKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));
    try {
      const secret = await getItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairLoadSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const saveKeypair = secret => {
  secret = chomp(secret);

  return async dispatch => {
    dispatch(keypairProcessing(true));

    if (!isValidSecret(secret)) {
      const error = new Error('Invalid Secret');
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }

    try {
      await setItem(STORAGE_KEY, secret);
      dispatch(keypairProcessing(false));
      dispatch(keypairSaveSucceded({ secret }));
      return Promise.resolve({ secret });
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};

export const removeKeypair = () => {
  return async dispatch => {
    dispatch(keypairProcessing(true));

    try {
      await removeItem(STORAGE_KEY);
      dispatch(keypairProcessing(false));
      dispatch(keypairRemoveSucceded());
      return Promise.resolve();
    } catch (error) {
      dispatch(keypairProcessing(false));
      dispatch(keypairErrored({ error }));
      return Promise.reject({ error });
    }
  };
};
