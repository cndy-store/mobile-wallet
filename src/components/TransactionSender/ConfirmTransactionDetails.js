import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  StatusBar,
  StyleSheet,
  View,
  Text,
  TextInput
} from 'react-native';
import Dimensions from 'Dimensions';
import ShortenedPublicKey from '../../components/ShortenedPublicKey';
import { sanitizeOngoingAmountInput } from '../../lib/formatter';

const { width, height } = Dimensions.get('window');

class ConfirmTransactionDetails extends Component {
  constructor(props) {
    super(props);

    this.handleConfirm = this.handleConfirm.bind(this);
    this.handleReject = this.handleReject.bind(this);
  }

  handleConfirm() {
    this.props.onConfirm();
  }

  handleReject() {
    this.props.onReject();
  }

  render() {
    return (
      <View>
        <ShortenedPublicKey publicKey={this.props.receiver} />
        <Text>Amount:</Text>
        <Text
          style={{
            width,
            height: 40
          }}
        >
          {this.props.amount}
        </Text>
        <Button title="Confirm!" onPress={this.handleConfirm} />
        <Button title="Change!" onPress={this.handleReject} />
      </View>
    );
  }
}

ConfirmTransactionDetails.propTypes = {
  receiver: PropTypes.string.isRequired,
  amount: PropTypes.string.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onReject: PropTypes.func.isRequired
};

export default ConfirmTransactionDetails;
