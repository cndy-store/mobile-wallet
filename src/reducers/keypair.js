import {
  KEYPAIR_STARTS_PROCESSING,
  KEYPAIR_ERROR,
  KEYPAIR_LOAD,
  KEYPAIR_SAVE,
  KEYPAIR_REMOVE
} from '../actions/keypair';

const defaultState = {
  isProcessing: false,
  error: null,
  keypair: null
};

export default function keypair(state = defaultState, action) {
  switch (action.type) {
    case KEYPAIR_STARTS_PROCESSING:
      return { ...state, isProcessing: action.isProcessing };
    case KEYPAIR_ERROR:
      return {
        ...state,
        error: action.error,
        isProcessing: action.isProcessing
      };
    case KEYPAIR_LOAD:
    case KEYPAIR_SAVE:
    case KEYPAIR_REMOVE:
      return {
        ...state,
        keypair: action.keypair,
        error: action.error,
        isProcessing: action.isProcessing
      };
    default:
      return state;
  }
}
