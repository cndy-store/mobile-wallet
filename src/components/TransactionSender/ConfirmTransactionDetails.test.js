import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from 'react-native';
import { publicKey } from '../../__tests__/fixtures/keypair';
import ConfirmTransactionDetails from './ConfirmTransactionDetails';

it('renders correctly without an error', () => {
  const onConfirm = jest.fn();
  const onReject = jest.fn();

  const tree = renderer
    .create(
      <ConfirmTransactionDetails
        receiver={publicKey}
        amount={'12.34'}
        onConfirm={onConfirm}
        onReject={onReject}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('calls the onConfirm prop if confirm button is pressed', () => {
  const onConfirm = jest.fn();
  const onReject = jest.fn();

  const instance = renderer.create(
    <ConfirmTransactionDetails
      receiver={publicKey}
      amount={'12.34'}
      onConfirm={onConfirm}
      onReject={onReject}
    />
  ).root;

  instance.findByProps({ title: 'Confirm!' }).props.onPress();
  expect(onConfirm).toHaveBeenCalled();
});

it('calls the onConfirm prop if reject button is pressed', () => {
  const onConfirm = jest.fn();
  const onReject = jest.fn();

  const instance = renderer.create(
    <ConfirmTransactionDetails
      receiver={publicKey}
      amount={'12.34'}
      onConfirm={onConfirm}
      onReject={onReject}
    />
  ).root;

  instance.findByProps({ title: 'Change!' }).props.onPress();
  expect(onReject).toHaveBeenCalled();
});
