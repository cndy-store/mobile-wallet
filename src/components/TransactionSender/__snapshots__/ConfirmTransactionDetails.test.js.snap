// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without an error 1`] = `
<View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="middle"
    numberOfLines={1}
    style={
      Object {
        "fontSize": 24,
      }
    }
  >
    GCHSUHOKNZQAKTULV5YTMG5CKU2NKE7UUQVN6JT2JY7337CRBIAZ6LXK
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Amount:
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "height": 40,
        "width": 750,
      }
    }
  >
    12.34
  </Text>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={undefined}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Confirm!
      </Text>
    </View>
  </View>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={undefined}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Change!
      </Text>
    </View>
  </View>
</View>
`;
